src/tasks/index.ts

```diff
@@ -118,10 +118,12 @@ export async function setup(): Promise<void> {
 }
 
 // Register a task handler
-export function addTaskHandler(
+export function addTaskHandler<T extends Schema>(
 	name: string,
-	fn: TaskHandler['fn'],
-	schema?: Schema,
+	fn: T extends undefined
+		? TaskHandler<undefined>['fn']
+		: TaskHandler<FromSchema<NonNullable<T>>>['fn'],
+	schema?: T,
```

Comment:
```suggestion
export function addTaskHandler<T extends Schema>(
	name: string,
	fn: TaskHandler<NonNullable<Task['Read']['is_executed_with__parameter_set']>>['fn'], // unknown would have been simpler perhaps :)
	schema?: undefined,
): void;
export function addTaskHandler<T extends Schema>(
	name: string,
	fn: TaskHandler<FromSchema<NonNullable<T>>>['fn'],
	schema: T,
): void;
export function addTaskHandler<T extends Schema>(
	name: string,
	fn: TaskHandler['fn'],
	schema?: T,
): void {
```

Reply:
I love it. thanks

Reply:
@thgreasi mayb better than using `Task['Read']['is_executed_with__parameter_set']` we could use just `{}` as this is whhat it will default to when no schema is present. The `NonNullable<Task['Read']['is_executed_with__parameter_set']> could also be an array. See: https://github.com/balena-io/pinejs/blob/c94bbc91395a1afbabf44f09053c96d813ab9d82/src/tasks/worker.ts#L117

Reply:
Afaict, tasks atm could work with arrays as a parameter set though, since the DB allows it on that field, we don't have any runtime check for arrays, and the ajv validation would work as far as you define your schema as an array. So I would lean towards sticking to what the DB says atm.